driver-core/src/main/com/mongodb/internal/connection/AwsAuthenticator.java

```diff
@@ -223,11 +224,19 @@ private String getPassword() {
 
         @Nullable
         private String getSessionToken() {
-            if (credential.getPassword() != null && credential.getUserName() != null) {
-                return credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
+            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
+            if (credential.getUserName() != null) {
+                return token;
+            }
+            if (token != null) {
+                throw new IllegalArgumentException("The connection string contains a session token without credentials");
             }
 
-            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null && System.getenv("AWS_ACCESS_KEY_ID") != null) {
+            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null || System.getenv("AWS_ACCESS_KEY_ID") != null) {
+                if (System.getenv("AWS_SECRET_ACCESS_KEY") == null || System.getenv("AWS_ACCESS_KEY_ID") == null) {
+                    throw new IllegalArgumentException("The environment variables 'AWS_ACCESS_KEY_ID' and 'AWS_SECRET_ACCESS_KEY' must "
```

Comment:
This change (to throw if only one of the two env vars is set seems like a spec change (I recall Mark saying server just silently ignores in this case).  What do you think?  Do we want this change?

Reply:
So I updated the Spec ([here](https://github.com/mongodb/specifications/pull/744/commits/7fb511615b09b4c75a19a3b45f66e5a0a18a5804)) to specify that we should error if we have missing credentials. This code assume that the user wants to authenticate a specific way if one of the three credentials is set and fail if we don't have both a user and password.